### PR TITLE
Kernel: Avoid panics in kmalloc()

### PR DESCRIPTION
**Pull request by arraystock**
Previously, whenever the system ran out of memory and a process attempted to allocate some, the kernel would simply panic.

Now the kernel simply crashes the process trying to allocate memory, unless the process is ring 0.

To test this, use the `allocate` program and try to allocate more memory than available. Before, the system would kp.

Let me know if I need to change anything! :)

### CODE REVIEW

File: Kernel/kmalloc.cpp [hunk 1]
```diff
@@ -107,9 +107,16 @@ void* kmalloc_impl(size_t size)
...
+        kprintf("%s(%u) kmalloc(): Uh-oh! Out of memory.\n", current->process().name().characters(), current->pid());
+        kprintf("sum_free=%u, real_size=%u\n", sum_free, real_size);
+        if (current->process().is_ring0()) {
+            kprintf("Out of memory: PANIC!\n");
+            dump_backtrace();
+            hang();
+        }
+        kprintf("Crashing active process %s(%u).\n", current->process().name().characters(), current->pid());
+        current->process().crash();
+        return {};
```
Comment (bugaevc) [inline Kernel/kmalloc.cpp]: This is unreachable, right? Why not make it `ASSERT_NOT_REACHED()` instead?

Comment (arraystock) [reply Kernel/kmalloc.cpp]: Alright, I've pushed a commit that changes the `return {};` to `ASSERT_NOT_REACHED()`.

**Comment by awesomekling** [comment]
Hi Jonathan, thank you for working on Serenity :)

I think this patch would take us in the wrong direction, so I'm going to reject it. Let me explain why:

The fact that `kmalloc()` never fails is a kernel invariant. (Although it certainly often does fail at the moment, which is why we panic in those cases.)
The kernel is not designed to handle a process suddenly crashing *inside the kernel* because in order to crash safely and cleanly, we would have to free any resources held by stack variables, release any locks taken, etc. This is not possible if we just cease execution of that process and throw away its stack memory.

I think there are two main ways to improve the situation:
- Fix whatever was causing kmalloc to run out of memory.
- Teach kmalloc to dynamically allocate more pages when it runs out (perhaps up to some limit.)

Also kmalloc is generally both crappy and slow, so building a better allocator that offers more tracing/profiling/etc might be a good idea too. Right now it's nontrivial to figure out "where all the kmalloc memory is going".

**Comment by arraystock** [comment]
Hi Andreas! I understand that completely. Your ideas are better than what I had came up with.

By the way, I think Serenity is seriously amazing, and I can’t wait to see where the future takes it :)

**Pull request closed without merging**